const-generic-const-fn-bounds.md

```diff
@@ -0,0 +1,499 @@
+- Feature Name: const_generic_const_fn_bounds
+- Start Date: 2018-10-05
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow `impl const Trait` for trait impls where all method impls are checked as const fn.
+
+Make it legal to declare trait bounds on generic parameters of const functions and allow
+the body of the const fn to call methods on the generic parameters that have a `const` modifier
+on their bound.
+
+# Motivation
+[motivation]: #motivation
+
+Currently one can declare const fns with generic parameters, but one cannot add trait bounds to these
+generic parameters. Thus one is not able to call methods on the generic parameters (or on objects of the
+generic parameter type), because they are fully unconstrained.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+You can call methods of generic parameters of a const function, because they are implicitly assumed to be
+`const fn`. For example, the `Add` trait bound can be used to call `Add::add` or `+` on the arguments
+with that bound.
+
+```rust
+const fn triple_add<T: Add>(a: T, b: T, c: T) -> T {
+    a + b + c
+}
+```
+
+The obligation is passed to the caller of your `triple_add` function to supply a type whose `Add` impl is fully
+`const`. Since `Add` only has `add` as a method, in this case one only needs to ensure that the `add` method is
+`const`. Instead of adding a `const` modifier to all methods of a trait impl, the modifier is added to the entire
+`impl` block:
+
+```rust
+struct MyInt(i8);
+impl const Add for MyInt {
+    fn add(self, other: Self) -> Self {
+        MyInt(self.0 + other.0)
+    }
+}
+```
+
+The const requirement is inferred on all bounds of the impl and its methods,
+so in the following `H` is required to have a const impl of `Hasher`, so that
+methods on `state` are callable.
+
+```rust
+impl const Hash for MyInt {
+    fn hash<H>(
+        &self,
+        state: &mut H,
+    )
+        where H: Hasher
+    {
+        state.write(&[self.0 as u8]);
+    }
+}
+```
+
+The same goes for associated types' bounds: all the bounds require `impl const`s for the type used
+for the associated type:
+
+```rust
+trait Foo {
+    type Bar: Add;
+}
+impl Foo for A {
+    type Bar = B; // B must have an `impl const Add for B`
+}
+```
+
+If an associated type has no bounds in the trait, there are no restrictions to what types may be used
+for it.
+
+These rules for associated types exist to make this RFC forward compatible with adding const default bodies
+for trait methods. These are further discussed in the "future work" section.
+
+## Generic `impl` blocks
+
+Similar to generic parameters on `const` functions, one can have generic parameters on `impl` blocks.
+These follow the same rules as bounds on `const` functions:
+
+* all bounds are required to have `impl const` for substituted types if the impl is used in a const context
+    * except in the presence of `?const` (see below)
```

Comment:
The fact that this exception is first mentioned when it comes to impl blocks makes it sound like it does not apply for what got previously discussed.

Reply:
I find the overall structure here a bit confusing.  You first have some text outside any subsection that seems to give an overview, cutting across a bunch of features, and then you start to systematically explore `const Trait` in some areas -- but there is no subsection specifically for "`const Trait` in function bounds".  Hence this is the first time you are summarizing, but you are summarizing for a *new* area that we didn't look at, and you are also adding a new concept while summarizing.  Now it is unclear how much of this summary also applies to the previously discussed area of bounds for functions.

Reply:
It's not supposed to be a summary, but just a terse explanation of the new feature. Due to the similarity to bounds on `const` functions I wanted to highlight the differences and similarities in a bullet point way.

Reply:
I think the structure would be improved by extending this subsection to cover generic bounds in both `fn` and `impl` contexts. The rules are the same, after all.